src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,12 +7608,791 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+    pOptBoolsDsc->cmpOp    = cmpOp;
+
+    optOptimizeBoolsUpdateTrees(pOptBoolsDsc);
+
+    // Update the change to true to continue the bool optimization for the rest of the BB chain
+    changeCond = true;
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("Folded %sboolean conditions of " FMT_BB " and " FMT_BB " to :\n", c2->OperIsLeaf() ? "" : "non-leaf ",
+               pOptBoolsDsc->b1->bbNum, pOptBoolsDsc->b2->bbNum);
+        gtDispStmt(s1);
+        printf("\n");
+    }
+#endif
+
+    return changeCond;
+}
+
+//
+// optOptimizeBoolsChkBlkCond: Checks block conditions if it can be boolean optimized
+//
+// Arguments:
+//      pOptBoolsDsc    The descriptor with Basic Block b1, b2 and b3 set when the method is called
+//
+// Return:
+//      s1              If all conditions pass, returns the last statement of b1.
+//
+// Notes:
+//      This method checks if the second or third block contains only one statement, and
+//      checks if tree operators are of the right type, e.g, GT_JTRUE, GT_RETURN.
+//
+//      When called, b1, b2 and b3 of pOptBoolsDsc are set, where
+//          b1 is the first block, b2 is the second block and b3 is the third block.
+//      If it passes all the conditions, t1, t2 and t3 are set to the root nodes of b1, b2 and b3 each.
+//      SameTarget is also updated to true if b1 and b2 jump to the same destination.
+//
+Statement* Compiler::optOptimizeBoolsChkBlkCond(OptBoolsDsc* pOptBoolsDsc)
+{
+    BasicBlock* b1 = pOptBoolsDsc->b1;
+    BasicBlock* b2 = pOptBoolsDsc->b2;
+    BasicBlock* b3 = pOptBoolsDsc->b3;
+
+    bool optReturnBlock = false;
+    if (b3)
+    {
+        optReturnBlock = true;
+    }
+
+    // Check jump destination condition
+
+    if (!optReturnBlock)
+    {
+        if (b1->bbJumpDest == b2->bbJumpDest)
+        {
+            // Given the following sequence of blocks :
+            //        B1: brtrue(t1, BX)
+            //        B2: brtrue(t2, BX)
+            //        B3:
+            // we will try to fold it to :
+            //        B1: brtrue(t1|t2, BX)
+            //        B3:
+
+            pOptBoolsDsc->sameTarget = true;
+        }
+        else if (b1->bbJumpDest == b2->bbNext)
+        {
+            // Given the following sequence of blocks :
+            //        B1: brtrue(t1, B3)
+            //        B2: brtrue(t2, BX)
+            //        B3:
+            // we will try to fold it to :
+            //        B1: brtrue((!t1)&&t2, BX)
+            //        B3:
+
+            pOptBoolsDsc->sameTarget = false;
+        }
+        else
+        {
+            return nullptr;
+        }
+    }
+    else
+    {
+        // Does b1 jump to b3?
+        // One example: Given the following sequence of blocks :
+        //        B1: brtrue(!t1, B3)
+        //        B2: return(t2)
+        //        B3: return(false)
+        //    we will try to fold it to :
+        //        B1: return(t1|t2)
+        if (b1->bbJumpDest != b2->bbNext)
```

Comment:
```suggestion
        if (b1->bbJumpDest != b3)
```

I suggested elsewhere that we shouldn't use `b2->bbNext` for `b2` of `BBJ_RETURN`, when considering flow.

Reply:
Done.